include/file.hpp

```diff
@@ -0,0 +1,205 @@
+				}
+			}
+
+			assert(!(mode & std::ios_base::app));
+			assert(!(mode & std::ios_base::trunc));
+			return true;
+		}
+		StdioWrapper *close() { return fclose(_file) == 0 ? this : nullptr; }
+
+		char const *name() const {
+			return _file == stdin ? "<stdin>" : "<stdout>";
+		}
+
+	protected:
+		// Called by `sgetc()`, since we don't use `gptr()`.
+		int_type underflow() override {
+			int_type c = uflow();
+			if (c != traits_type::eof()) {
+				// This function must not advance the read pointer, so put the character back.
+				int ret = ungetc(c, _file);
+				// Since we just read a character, pushback should be guaranteed.
+				assert(ret == c);
+			}
+			return c;
+		}
+		// Called by `sbumpc()`, since we don't use `gptr()`.
+		int_type uflow() override {
+			static_assert(traits_type::eof() == EOF, "Error values do not match");
+			return fgetc(_file);
+		}
+		// Called by `sgetn()`.
+		std::streamsize xsgetn(char_type *s, std::streamsize count) override {
+			return fread(s, sizeof(char_type), count, _file);
+		}
+
+		// Called by `sputc()`, since we don't use `pptr()`.
+		int_type overflow(int_type c) override { return fputc(c, _file); }
+		// Called by `sputn()`.
+		std::streamsize xsputn(char_type const *s, std::streamsize count) override {
+			return fwrite(s, sizeof(char_type), count, _file);
+		}
+
+		// Called by `pubsync()`.
+		int sync() override { return fflush(_file); }
+		// Called by `seekoff()`.
+		pos_type seekoff(off_type off, std::ios_base::seekdir dir,
+		                 std::ios_base::openmode which = std::ios_base::in
+		                                                 | std::ios_base::out) override {
+			if (_file == stdin) {
+				assert(which & std::ios_base::in);
+			} else {
+				assert(_file == stdout);
+				assert(which & std::ios_base::out);
+			}
+
+			switch (dir) {
+			case std::ios_base::beg:
+				return fseek(_file, off, SEEK_SET);
+			case std::ios_base::end:
+				return fseek(_file, off, SEEK_END);
+			case std::ios_base::cur:
+				return fseek(_file, off, SEEK_CUR);
+			default:
+				unreachable_();
+			}
+		}
+	};
+
+public:
+	using char_type = char;
+	using traits_type = std::char_traits<char_type>;
+	using int_type = traits_type::int_type;
+	using pos_type = traits_type::pos_type;
+	using off_type = traits_type::off_type;
+
+private:
+	// Construct a `StdioWrapper` by default, since it's probably lighter than a `filebuf`.
+	std::variant<StdioWrapper, std::filebuf> _file;
+
+public:
+	File() : _file() {}
+	/**
+	 * This should only be called once, and before doing any `->` operations.
+	 * Returns `nullptr` on error, and a non-null pointer otherwise.
+	 */
+	File *open(std::filesystem::path const &path, std::ios_base::openmode mode) {
+		if (path == "-") {
+			return std::get<StdioWrapper>(_file).open(mode) ? this : nullptr;
+		} else {
+			// Replace the wrapper with an actual filebuf, and open the file.
+			return _file.emplace<std::filebuf>().open(path, mode) ? this : nullptr;
+		}
+	}
```

Comment:
Does this whole wrapper class just exist so that `open` can end up calling either `get(_file)` or `_file.emplace()`? I'm surprised there isn't a shorter stdlib-only solution.

Reply:
AIUI, the stdlib-only solution is `<ostream>`... which apparently we want to avoid? :P

(That said, now that I'm thinking about it, `ostream` has a `rdbuf` member function, which may actually provide access to a `filebuf` wrapper to `stdin`/`stdout`?)

Reply:
(Note that the wrapper class also exists as a smart-pointer, RAII owner for the `streambuf` implementation.)

Reply:
Just getting ahold of `cout`/`cin` requires pulling in `<iostream>`, which may be the very thing we're trying to avoid. Thoughts?

Reply:
Were we ever trying to avoid upstream entirely? Or just avoid its slow overloaded shift-operator API for reading/writing data?

Reply:
I know `#include <iostream>` incurs construction of the objects, and that much cannot be avoided with that solution. The shift-operator API *is* avoided like this, but I'm not sure it's slow because of the overloading?

Reply:
Personally I'd prefer simpler code using cout/cin, unless actual measured performance is significantly better by not calling their constructors and using file.hpp instead. (For the small files rgbgfx works with I doubt it matters.)

Reply:
Oh, `cin`/`cout` are only used to immediately pulling their `.rdbuf()` out of them.

Reply:
Right, but as you said, just including and constructing them adds overhead (<https://www.ibm.com/support/pages/apar/PJ42715>), although it might be minimal.

Reply:
In the mid-2000s, I saw executable size overhead on the order of 175 kB of overhead just from constructing one of those objects. At the time, both MinGW and devkitARM were using statically linked GNU libstdc++. The rabbit hole of where all the space was going to led to the word `moneypunct`, and I concluded that the constructor of `cout` was constructing members for formatting date, time, and currency objects just in case the program might print one of those. This brought in the whole locale library as ultimately unreachable code that I couldn't exclude from the executable even with `-Wl,--gc-sections`.

Reply:
On my machine (using Cygwin `g++` 11.3.0), `rgbgfx` just grows from 223 KB to 226 KB when I trivially use them (`#include <iostream> std::string name; std::cin >> name; std::cout << "hello, " << name << "!\n";`).

Reply:
I'm personally more concerned about the speed impact, but it seems negligible.